frostsnap_coordinator/src/backup_run.rs

```diff
@@ -0,0 +1,160 @@
+    }
+}
+
+#[derive(Default, Debug)]
+pub struct BackupState {
+    pub runs: BTreeMap<KeyId, BackupRun>,
+}
+
+#[derive(Debug, Clone)]
+pub enum BackupMutation {
+    StartBackup {
+        key_id: KeyId,
+        devices: Vec<DeviceId>,
+    },
+    MarkDeviceComplete {
+        key_id: KeyId,
+        device_id: DeviceId,
+        timestamp: u32,
+    },
+    MarkRunComplete {
+        key_id: KeyId,
+        timestamp: u32,
+    },
+}
+
+// We use device_id := "_completed" as a filler to timestamp the last completed backup run for a key
+impl Persist<rusqlite::Connection> for BackupState {
+    type Update = Vec<BackupMutation>;
+    type InitParams = ();
+
+    fn initialize(conn: &mut rusqlite::Connection, _: ()) -> anyhow::Result<Self> {
+        conn.execute(
+            "CREATE TABLE IF NOT EXISTS backup_runs (
+                key_id TEXT NOT NULL,
+                device_id TEXT NOT NULL,
+                timestamp INTEGER,
+                PRIMARY KEY (key_id, device_id)
+            )",
+            [],
+        )?;
+
+        let mut stmt = conn.prepare(
+            "SELECT key_id, device_id, timestamp 
+             FROM backup_runs 
+             WHERE device_id != '_completed'
+             ORDER BY key_id",
+        )?;
+
+        let rows = stmt.query_map([], |row| {
+            Ok((
+                row.get::<_, String>(0)?,      // key_id
+                row.get::<_, String>(1)?,      // device_id
+                row.get::<_, Option<u32>>(2)?, // timestamp
+            ))
+        })?;
+
+        // group by key_id and build BackupRun for each group
+        let runs = rows
+            .into_iter()
+            .collect::<Result<Vec<_>, _>>()?
+            .into_iter()
+            .fold(
+                BTreeMap::new(),
+                |mut map: BTreeMap<_, Vec<_>>, (key_id, device_id, timestamp)| {
+                    map.entry(key_id).or_default().push((device_id, timestamp));
+                    map
+                },
+            )
+            .into_iter()
+            .map(|(key_id, devices)| {
+                let devices = devices
+                    .into_iter()
+                    .map(|(device_id, timestamp)| Ok((device_id.parse()?, timestamp)))
+                    .collect::<Result<Vec<_>, anyhow::Error>>()?;
+
+                Ok((key_id.parse()?, BackupRun { devices }))
+            })
+            .collect::<Result<BTreeMap<_, _>, anyhow::Error>>()?;
+
+        Ok(BackupState { runs })
+    }
+
+    fn persist_update(
+        conn: &mut rusqlite::Connection,
+        update: Vec<BackupMutation>,
+    ) -> anyhow::Result<()> {
+        for mutation in update {
+            match mutation {
+                BackupMutation::StartBackup { key_id, devices } => {
+                    // Delete all except completion record
+                    conn.execute(
+                        "DELETE FROM backup_runs WHERE key_id = ?1 AND device_id != '_completed'",
```

Comment:
It seems like you are trying to keep track of which backup has been completed. But why can't you just tell this by how many devices have been completed?

It is very weird to insert something into the `device_id` column that is not a `device_id`.

Reply:
The idea was to store some knowledge that a backup run has been completed before.

If a user starts a subsequent backup run, I was thinking that we should let them cancel it (maybe mid way though) and get rid of the warning banner. I haven't implemented this yet though.

Maybe a bit too hacky? Was avoiding another separate table to store a history of completed backup runs.

Reply:
Do we need a history? As far as I understand there is a single backup task with each wallet. It can be in a state of `None` or `Some(BackupProgress)`. I don't think it needs to be any more complicated than that. We don't need to keep around whether they have completed one before this one. Just let them cancel it and set it back to `None`.